File: flutter_client/test/ticket_qty_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_client/features/trade/logic/ticket_qty.dart';

void main() {
  test('CN buy 100% rounds down to board lot', () {
    expect(
      ticketQtyForPercent(
        percent: 100,
        side: 'buy',
        market: 'CN',
        price: 10,
        cash: 10000,
      ),
      1000,
    );
  });

  test('CN buy 25% rounds down to 100-share lot', () {
    // 10000/10 = 1000 股，25% = 250 → 200。
    expect(
      ticketQtyForPercent(
        percent: 25,
        side: 'buy',
        market: 'CN',
        price: 10,
        cash: 10000,
      ),
      200,
    );
  });

  test('US buy 25% uses 1-share lot', () {
    // 10000/190 ≈ 52 股，25% = 13。
    expect(
      ticketQtyForPercent(
        percent: 25,
        side: 'buy',
        market: 'US',
        price: 190,
        cash: 10000,
      ),
      13,
    );
  });

  test('sell 50% US', () {
    expect(
      ticketQtyForPercent(
        percent: 50,
        side: 'sell',
        market: 'US',
        sellable: 300,
      ),
      150,
    );
  });

  test('sell CN 50% rounds down to lot', () {
    expect(
      ticketQtyForPercent(
        percent: 50,
        side: 'sell',
        market: 'CN',
        sellable: 250,
      ),
      100,
    );
  });

  test('missing price or cash yields 0', () {
    expect(
      ticketQtyForPercent(
        percent: 100,
        side: 'buy',
        market: 'US',
        price: null,
        cash: 10000,
      ),
      0,
    );
    expect(
      ticketQtyForPercent(
        percent: 100,
        side: 'buy',
        market: 'US',
        price: 10,
        cash: 0,
      ),
      0,
    );
  });

  test('lotSize and cash currency', () {
    expect(lotSizeForMarket('CN'), 100);
    expect(lotSizeForMarket('US'), 1);
    expect(cashCurrencyForMarket('HK'), 'HKD');
  });
}

File: ruoyi-fastapi-frontend/src/mobile/utils/ticketQty.test.js
import { describe, it } from 'node:test'
import assert from 'node:assert/strict'
import { lotSizeForMarket, ticketQtyForPercent, inferMarket } from './ticketQty.js'

describe('lotSizeForMarket', () => {
  it('uses 100 for CN and 1 for HK/US', () => {
    assert.equal(lotSizeForMarket('CN'), 100)
    assert.equal(lotSizeForMarket('HK'), 1)
    assert.equal(lotSizeForMarket('US'), 1)
  })
})

describe('ticketQtyForPercent', () => {
  it('returns 0 when buy price or cash is missing', () => {
    assert.equal(ticketQtyForPercent({ percent: 100, side: 'buy', market: 'US', price: null, cash: 10000 }), 0)
    assert.equal(ticketQtyForPercent({ percent: 100, side: 'buy', market: 'US', price: 10, cash: null }), 0)
    assert.equal(ticketQtyForPercent({ percent: 100, side: 'buy', market: 'US', price: 0, cash: 10000 }), 0)
  })

  it('floors US buy lots to whole shares', () => {
    assert.equal(ticketQtyForPercent({ percent: 100, side: 'buy', market: 'US', price: 10, cash: 105 }), 10)
    assert.equal(ticketQtyForPercent({ percent: 50, side: 'buy', market: 'US', price: 10, cash: 100 }), 5)
  })

  it('floors CN buy lots to 100', () => {
    assert.equal(ticketQtyForPercent({ percent: 100, side: 'buy', market: 'CN', price: 10, cash: 2500 }), 200)
    assert.equal(ticketQtyForPercent({ percent: 25, side: 'buy', market: 'CN', price: 10, cash: 10000 }), 200)
    assert.equal(ticketQtyForPercent({ percent: 100, side: 'buy', market: 'CN', price: 10, cash: 99 }), 0)
  })

  it('sells A-share in lots of 100 and never invents qty', () => {
    assert.equal(ticketQtyForPercent({ percent: 100, side: 'sell', market: 'CN', sellable: 350 }), 300)
    assert.equal(ticketQtyForPercent({ percent: 50, side: 'sell', market: 'CN', sellable: 350 }), 100)
    assert.equal(ticketQtyForPercent({ percent: 100, side: 'sell', market: 'CN', sellable: 0 }), 0)
    assert.equal(ticketQtyForPercent({ percent: 100, side: 'sell', market: 'US', sellable: 7 }), 7)
  })
})

describe('inferMarket', () => {
  it('reads suffix and numeric codes', () => {
    assert.equal(inferMarket('0700.HK'), 'HK')
    assert.equal(inferMarket('600519'), 'CN')
    assert.equal(inferMarket('AAPL'), 'US')
  })
})
